### PR TITLE
[Merged by Bors] - Remove Bytes, FromBytes, Labels, EntityLabels. Document rest of bevy_core and enable warning on missing docs.

### PR DESCRIPTION
**Pull request by james7132**
This PR is part of the issue #3492.

# Objective

 - Clean up dead code in `bevy_core`.
 - Add and update the `bevy_core` documentation to achieve a 100% documentation coverage.
 - Add the #![warn(missing_docs)] lint to keep the documentation coverage for the future.

# Solution

 - Remove unused `Bytes`, `FromBytes`, `Labels`, and `EntityLabels` types and associated systems.
 - Made several types private that really only have use as internal types, mostly pertaining to fixed timestep execution.
 - Add and update the bevy_core documentation.
 - Add the #![warn(missing_docs)] lint.

# Open Questions

Should more of the internal states of `FixedTimestep` be public? Seems mostly to be an implementation detail unless someone really needs that fixed timestep state.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> Should more of the internal states of FixedTimestep be public? Seems mostly to be an implementation detail unless someone really needs that fixed timestep state.

I'll defer to @maniwani (Joy) here; they've thought most deeply about this.

**Comment by james7132** [comment]
Does this PR require a migration guide? Not sure how many users were dependent on those public types, including the ones made private.

**Comment by alice-i-cecile** [comment]
> Does this PR require a migration guide? Not sure how many users were dependent on those public types, including the ones made private.


Yep, we'll toss it in the tag. It's not critical, but it is nice to have *some* indication that we made breaking changes.

File: crates/bevy_core/src/name.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@ use std::{
 };
 
 /// Component used to identify an entity. Stores a hash for faster comparisons
```
Comment (alice-i-cecile) [inline crates/bevy_core/src/name.rs:10]: I think this needs a comment explaining that names are not guaranteed to be globally unique. And probably also that `Entity` should be used as the default "unique identifier" instead.

Especially with `Hash` being implemented, that's an easy and painful mistake to make.

File: crates/bevy_core/src/time/fixed_timestep.rs [hunk 2]
```diff
@@ -10,45 +10,65 @@ use bevy_ecs::{
...
+/// This does not guarentee that the time elapsed between executions is exactly the provided
+/// fixed timestep, but will guarentee that the execution will run multiple times per game tick
+/// until the number of repetitions is as expected.
+///
+/// For example, a system with a fixed timestep run criteria of 120 times per second will run
+/// two times during a ~16.667ms frame, once during a ~8.333ms frame, and once every two frames
+/// with ~4.167ms frames. However, the same criteria may not result in exactly 8.333ms passing
+/// between each execution.
+///
+/// When using this run criteria, it is advised not to rely on [`Time::delta`] or any of it's
```
Comment (alice-i-cecile) [inline crates/bevy_core/src/time/fixed_timestep.rs:67]: This should have another sentence explaining what users should do instead.

**Comment by alice-i-cecile** [comment]
Looks great; only a couple things to expand on.

**Comment by james7132** [comment]
Migration Guide for this PR:

# `Bytes`/`FromBytes` removed

The traits `Bytes` and `FromBytes` have been removed. These types were largely used for converting types to byte buffers for loading into rendering pipelines and some forms of serialization. With the render rework, `bevy_crevice` has largely taken over for the role of converting types to rendering friendly byte buffers. 

For serialization, both traits did not strictly define a binary representation or serialization format, and only allowed one representation per type. For these use cases, it's suggested to either use `serde` and a binary serialization format (i.e. `bincode`).

For conversion of plain-old-data (POD) types to their byte slice representation, it's suggested to directly use `bytemuck`'s functions instead to reinterpret said data.

# `Labels`/`EntityLabels` removed.

The component `Labels` and resource `EntityLabels` have been removed. These were largely used to label entities with string labels, and find all entities with a given label. These were added mostly in preparation for use in a editor.

It is suggested to instead use ZST (zero-size type) marker components, and add them to entities instead. This does not require runtime string hashing, and can be efficiently queried via queries as system parameters.

**Comment by alice-i-cecile** [comment]
You should also note that some of the state for `FixedTimeStep` was made private.

**Review by alice-i-cecile — APPROVED** [review]

**Review by mockersf — APPROVED** [review]
I'll be sad to see `Labels` go away, but it's the right thing to do 👍

**Comment by james7132** [comment]
> You should also note that some of the state for `FixedTimeStep` was made private.

Only `LocalFixedTimestepState` was privated and that doesn't look to be available in 0.5, looks like it was introduced after the fact. Added a note about the field of `FixedTimestepState` being privated, since those shouldn't be mutable outside of the crate.

**Pull request closed without merging**